scripts/base/protocols/redis/main.zeek

```diff
@@ -0,0 +1,254 @@
+@load base/protocols/conn/removal-hooks
+@load base/frameworks/signatures
+
+@load ./spicy-decls
+
+module Redis;
+
+export {
+	## Log stream identifier.
+	redef enum Log::ID += { LOG };
+
+	## The ports to register Redis for.
+	const ports = { 6379/tcp } &redef;
+
+	## Record type containing the column fields of the Redis log.
+	type Info: record {
+		## Timestamp for when the activity happened.
+		ts: time &log;
+		## Unique ID for the connection.
+		uid: string &log;
+		## The connection's 4-tuple of endpoint addresses/ports.
+		id: conn_id &log;
+		## The Redis command.
+		cmd: Command &log;
+		## The response for the command.
+		response: ServerData &log &optional;
+	};
+
+	## A default logging policy hook for the stream.
+	global log_policy: Log::PolicyHook;
+
+	global finalize_redis: Conn::RemovalHook;
+
+	type State: record {
+		## Pending requests.
+		pending: table[count] of Info;
+		## Current request in the pending queue.
+		current_request: count &default=0;
+		## Current response in the pending queue.
+		current_response: count &default=0;
+		## Ranges where we do not expect a response.
+		## Each range is one or two elements, one meaning it's unbounded, two meaning
+		## it begins at one and ends at the second.
+		no_response_ranges: vector of vector of count;
+		## We store if this analyzer had a violation to avoid logging if so.
+		## This should not be super necessary, but worth a shot.
+		violation: bool &default=F;
+	};
+
+	# Redis specifically mentions 10k commands as a good pipelining threshold, so
+	# we'll piggyback on that.
+	option max_pending_requests = 10000;
+}
+
+redef record connection += {
+	redis: Info &optional;
+	redis_state: State &optional;
+};
+
+redef likely_server_ports += { ports };
+
+event zeek_init() &priority=5
+	{
+	Log::create_stream(Redis::LOG, [ $columns=Info, $path="redis",
+	    $policy=log_policy ]);
+
+	Analyzer::register_for_ports(Analyzer::ANALYZER_SPICY_REDIS, ports);
+	}
+
+event analyzer_violation_info(atype: AllAnalyzers::Tag,
+    info: AnalyzerViolationInfo)
+	{
+	if ( atype == Analyzer::ANALYZER_SPICY_REDIS )
+		{
+		if ( info?$c )
+			{
+			if ( info$c?$redis_state )
+				{
+				info$c$redis_state$violation = T;
+				}
+			}
+		}
+	}
+
+function new_redis_session(c: connection): Info
```

Comment:
An `Info` record is more like a request/response transaction, no? "Session" to me would suggest multiple transactions.